crates/wasm-mutate/src/mutators/peephole/eggsy/mod.rs

```diff
@@ -0,0 +1,284 @@
+        costs: &mut HashMap<Id, (CF::Cost, usize)>,
+        eclass: &EClass<L, Option<i32>>,
+    ) -> Option<(CF::Cost, usize)> {
+        let (cost, node_idx) = eclass
+            .iter()
+            .enumerate()
+            .map(|(i, n)| (self.node_total_cost(n, costs), i))
+            .min_by(|a, b| cmp(&a.0, &b.0))
+            .unwrap_or_else(|| panic!("Can't extract, eclass is empty: {:#?}", eclass));
+        cost.map(|c| (c, node_idx))
+    }
+
+    fn node_total_cost(
+        &mut self,
+        node: &L,
+        costs: &mut HashMap<Id, (CF::Cost, usize)>,
+    ) -> Option<CF::Cost> {
+        let egraph = self.egraph;
+        let has_cost = |&id| costs.contains_key(&egraph.find(id));
+        if node.children().iter().all(has_cost) {
+            let costs = &costs;
+            let cost_f = |id| costs[&egraph.find(id)].0.clone();
+            Some(self.cost_function.cost(&node, cost_f))
+        } else {
+            None
+        }
+    }
+
+    pub fn get_costs(&self) -> &HashMap<Id, (<CF as CostFunction<L>>::Cost, usize)> {
+        &self.costs
+    }
+
+    /// Here is where the random magic is done
+    pub fn generate_random_tree(
+        &self,
+        rnd: &mut rand::prelude::SmallRng,
+        eclass: Id,
+        max_depth: u32,
+    ) -> crate::Result<(Vec<&L>, Vec<Vec<Id>>)> // return the random tree, TODO, improve the way the tree is returned
+    {
+        // A map from a node's id to its actual node data.
+        let mut id_to_node = vec![];
+        // A map from a parent node id to its child operand node ids.
+        let mut operands = vec![];
+
+        let egraph = self.egraph;
+        // Select a random node in this e-class
+        let rootidx = rnd.gen_range(0, egraph[eclass].nodes.len());
+        let rootnode = &egraph[eclass].nodes[rootidx];
+
+        id_to_node.push(&egraph[eclass].nodes[rootidx]);
+        operands.push(vec![]);
+
+        let mut worklist: Vec<_> = rootnode
+            .children()
+            .iter()
+            .map(|id| (eclass, 0, id, 0)) // (root, operant, depth)
+            .collect();
+
+        while let Some((parent, parentidx, &node, depth)) = worklist.pop() {
+            //println!("{:?}", parent);
+
+            let node_idx = if depth >= max_depth {
+                // look nearest leaf path, in this case, the best in AST size
+                self.costs[&node].1
+            } else {
+                rnd.gen_range(0, egraph[node].nodes.len())
+            };
+
+            //println!("random e-node {:?}", node_idx);
+            //println!("options {:?}", egraph[node].nodes);
+
+            let operand = Id::from(id_to_node.len());
+            let operandidx = id_to_node.len();
+            let last_node_id = parentidx; // id_to_node.len() - 1;
+            id_to_node.push(&egraph[node].nodes[node_idx]);
+            operands.push(vec![]);
+
+            operands[last_node_id].push(operand);
+
+            //let operand = &egraph[node].nodes[node_idx];
```

Comment:
Reminder to clean up this stuff and the commented out `println!`s above.

Reply:
Gotcha !